pyo3-derive-backend/src/pyenum.rs

```diff
@@ -0,0 +1,214 @@
+    enum_: &syn::Ident,
+    variants: &Vec<(syn::Ident, syn::ExprLit)>,
+) -> syn::Result<TokenStream> {
+    let matches = variants
+        .iter()
+        .map(|(ident, _)| {
+            variant_enumname(enum_, ident).map(|cls| {
+                quote! {
+                    #enum_::#ident => <#cls as pyo3::type_object::PyTypeObject>::type_object(py).to_object(py),
+                }
+            })
+        })
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    Ok(quote! {
+        impl pyo3::FromPy<#enum_> for pyo3::PyObject {
+            fn from_py(v: #enum_, py: pyo3::Python) -> Self {
+                match v {
+                    #(
+                        #matches
+                    )*
+                }
+            }
+        }
+    })
+}
+
+fn variant_enumname(enum_: &syn::Ident, cls: &syn::Ident) -> syn::Result<syn::Ident> {
+    let name = format!("{}_EnumVariant_{}", enum_, cls);
+    syn::parse_str(&name)
+}
+
+fn impl_const(enum_: &syn::Ident, cls: &syn::Ident) -> syn::Result<TokenStream> {
+    Ok(quote! {
+        #[classattr]
+        const #cls: #enum_ = #enum_::#cls;
+    })
+}
+
+fn impl_class(cls: &syn::Ident, parent: Option<&syn::Ident>) -> syn::Result<TokenStream> {
+    let (typ, desc) = if let Some(p) = parent {
+        (
+            variant_enumname(p, cls)?,
+            format!("variant {} of enum {}", cls, p),
+        )
+    } else {
+        // Clone is just to align the types
+        (cls.clone(), format!("enum {}", cls))
+    };
+
+    let clsname = cls.to_string();
+    let inventory = impl_methods_inventory(&typ);
+    let extractext = impl_extractext(&typ);
+    let protoregistry = impl_proto_registry(&typ);
+
+    Ok(quote! {
+        unsafe impl pyo3::type_object::PyTypeInfo for #typ {
+            type Type = #typ;
+            type BaseType = pyo3::PyAny;
+            type Layout = pyo3::PyCell<Self>;
+            type BaseLayout = pyo3::pycell::PyCellBase<pyo3::PyAny>;
+
+            type Initializer = pyo3::pyclass_init::PyClassInitializer<Self>;
+            type AsRefTarget = pyo3::PyCell<Self>;
+
+            const NAME: &'static str = #clsname;
+            const MODULE: Option<&'static str> = None;
+            const DESCRIPTION: &'static str = #desc;
+            const FLAGS: usize = 0;
+
+            #[inline]
+            fn type_object_raw(py: pyo3::Python) -> *mut pyo3::ffi::PyTypeObject {
+                use pyo3::type_object::LazyStaticType;
+                static TYPE_OBJECT: LazyStaticType = LazyStaticType::new();
+                TYPE_OBJECT.get_or_init::<Self>(py)
+            }
+        }
+
+        impl pyo3::PyClass for #typ {
+            type Dict =  pyo3::pyclass_slots::PyClassDummySlot ;
+            type WeakRef = pyo3::pyclass_slots::PyClassDummySlot;
+            type BaseNativeType = pyo3::PyAny;
+        }
+
+        #protoregistry
+
+        #extractext
+
+        impl pyo3::pyclass::PyClassAlloc for #typ {}
+
+        // TODO: handle not in send
```

Comment:
I think it's reasonable to expect `#[pyenum]` to always be `Send`? It wouldn't ever carry nontrivial data afaik.

Reply:
Maybe go further and have pyenum always be 'Copy' as we are doing small data here rather than ADTs?

Reply:
Perhaps, though I'm not sure it's our place to add `#[derive(Copy, Clone)]` automatically to the struct.